### ISSUE TITLE
suggestion: support area-under-PR-curve as eval metric for early stopping

### ISSUE THREAD

**Issue by kodonnell** [dmlc/xgboost#3054]
[Apparently](https://www.kaggle.com/general/7517) the area under the PR curve (what scikit-learn calls the 'average precision score') is a better metric for training unbalanced binary classification problems than the area under the ROC curve (i.e. 'auc'). How hard would this be to implement as an `eval_metric` when using earling stopping? I'm assuming implementing the calculation is simple, but it might require updating docs etc.

Alternatively, can we allow python users to provide their own scoring function (e.g. from `sklearn.metrics`)?

**Comment by kodonnell** [comment]
> Alternatively, can we allow python users to provide their own scoring function (e.g. from sklearn.metrics)?

Yes:

```python
def pr_auc_metric(y_predicted, y_true):
    return 'pr_auc', -skmetrics.average_precision_score(y_true.get_label(), y_predicted)
 ```

**Comment by data-steve** [comment]
Why is there a negative sign on `average_precision_score`? I thought average precision was positively correlated with improved performance, ie, bigger positive number is better?

**Comment by kodonnell** [comment]
As per [the docs](http://xgboost.readthedocs.io/en/latest/python/python_api.html?highlight=eval_metric#xgboost.XGBClassifier.fit):

> eval_metric (str, callable, optional) – ... This objective is always minimized.
